File: code/nn/neuron.go
package nn

import (
    "math/rand"
    "math"
)

const (
    ETA float64 = 0.15 // training rate
    ALPHA float64 = 0.5 // training momentum
)

type Connection struct {
    weight float64
    delta_weight float64
}

type Neuron struct {
    output float64
    output_weights []Connection
    index int
    gradient float64
}

type Layer []Neuron

func (n Neuron) UpdateInputWeights(previous_layer *Layer) {
    for _, neuron := range *previous_layer {
        old_delta_weight := neuron.output_weights[n.index].delta_weight

        new_delta_weight :=
            // Individual input, magnified by gradient and train rate
            ETA * neuron.output * n.gradient +
            // Also add momentum (fraction of previous delta weight)
            ALPHA * old_delta_weight

        neuron.output_weights[n.index].delta_weight = new_delta_weight
        neuron.output_weights[n.index].weight += new_delta_weight
    }
}

func (n Neuron) SumDOW(next_layer *Layer) float64 {
    sum := 0.0
    for i := 0; i < len(*next_layer) - 1; i++ {
        sum += n.output_weights[i].weight * (*next_layer)[i].gradient
    }
    return sum
}

func (n Neuron) CalculateHiddenGradients(next_layer *Layer) {
    dow := n.SumDOW(next_layer)
    n.gradient = dow * n.ActivationDerivative(n.output)
}
func (n Neuron) CalculateOutputGradients(target float64) {
    delta := target - n.output
    n.gradient = delta * n.ActivationDerivative(n.output)
}

func (n Neuron) Activation(x float64) float64 {
    return math.Tanh(x);
}
func (n Neuron) ActivationDerivative(x float64) float64 {
    // Approximation of derivative of tanh
    return 1.0 - x*x
}

// TODO: Not an apt name
func (n Neuron) FeedForward(previous_layer *Layer) {
    sum := 0.0

    // Sum outputs from the previous layer.
    for _, neuron := range *previous_layer {
        // TODO: there may be duplication issues here
        sum += neuron.output * neuron.output_weights[n.index].weight
    }

    n.output = n.Activation(sum)
}

func (n Neuron) New(num_outputs int, index int) {
    for c := 0; c < num_outputs; c++ {
        n.output_weights = append(n.output_weights, Connection{weight: rand.Float64()})
    }
    n.index = index
}

File: code/neuron.cpp
#include <vector>
#include <cmath>
#include "neuron.h"

#define ETA 0.05 // Learning rate [0..1]
#define ALPHA 0.5 // Momentum (multiplier of last change in weight) [0..]

using namespace std;


void Neuron::update_input_weights(Layer &previous_layer) {
	// The weights to be updated are in the Connection container
	// in the nuerons in the preceding layer

	for (int n = 0; n < previous_layer.size(); ++n) {
		Neuron &neuron = previous_layer[n];
		double old_delta_weight = neuron.output_weights[this->index].delta_weight;

		double new_delta_weight =
				// Individual input, magnified by the gradient and train rate:
				ETA
				* neuron.get_output()
				* this->gradient
				// Also add momentum, a fraction of the previous delta weight
				+ ALPHA
				* old_delta_weight;
		neuron.output_weights[this->index].delta_weight = new_delta_weight;
		neuron.output_weights[this->index].weight += new_delta_weight;
	}
}
double Neuron::sum_dow(const Layer &next_layer) const {
	double sum = 0.0;

	// Sum our contributions of the errors at the nodes we feed

	for (int n = 0; n < next_layer.size() - 1; ++n) {
		sum += this->output_weights[n].weight * next_layer[n].gradient;
	}

	return sum;
}

void Neuron::calculate_hidden_gradients(const Layer &next_layer) {
	double dow = sum_dow(next_layer);
	this->gradient = dow * Neuron::activation_derivative(this->output);
}
void Neuron::calculate_output_gradients(double targets) {
	double delta = targets - this->output;
	this->gradient = delta * Neuron::activation_derivative(this->output);
}

double Neuron::activation(double x) {
	// Bring output into [-1..1] range
	return tanh(x);
}

double Neuron::activation_derivative(double x) {
	// Approximation of tanh derivative
	return 1.0 - x * x;
}

void Neuron::feed_forward(const Layer &previous_layer) {
	double sum = 0.0;

	// Sum outputs from the previous layer.
	for (int n = 0; n < previous_layer.size(); ++n)
		sum += previous_layer[n].get_output() * previous_layer[n].output_weights[this->index].weight;

	this->output = Neuron::activation(sum);
}

Neuron::Neuron(int num_outputs, int index) {
	for (int c = 0; c < num_outputs; ++c){
		Connection connection;
		connection.weight = 1.0/2.0;
		this->output_weights.push_back(connection);
	}

	this->index = index;
}
